cEP-0005.md

```diff
@@ -0,0 +1,157 @@
+ignore += ./vendor
+
+max_line_length = 80
+
+# This inherits all settings from the `all` section. There is no implicit
+# inheritance.
+[all.python]
+language = Python
+files = **.py
+aspects = smell, redundancy.clone  # Inclusive
+
+[all.c]
+language = C
+files = **.(c|h)
+ignore_aspects = redundancy  # Exclusive
+```
+
+### Bear API
+
+Bears will receive the aspects that should be checked as a parameter and
+results will get metadata to indicate what aspect has been checked.
+
+```python
+from ... import LocalBear
+from ... import Result
+from ...Aspects import Redundancy
+
+
+class RedundancyBear(LocalBear):
+    # coala will run bears based on this metadata
+    LANGUAGES = {"C", "Python"}
+    ASPECTS = {Redundancy.Clone, Redundancy.UnusedImport}
+
+    # Aspects are passed as parameter
+    def run(self, filename, file, aspects):
+        # No documentation required anymore for the bears and their parameters.
+
+        # Check the code ...
+
+        # A bear may yield results for any aspect even if it's not selected but
+        # may use the selected aspects to save performance. coala will filter
+        # out only selected aspects.
+        yield Result.from_values(aspect=Redundancy.Clone, ...)
+```
+
+The concept of aspects allows us to implement a consistency check and reduce
+documentation redundancy. Instead of documenting settings and results in every
+bear, we can create a new aspects class:
+
+```python
+class Redundancy(Aspect):
+    """
+    This meta aspect describes any kind of redundancy in your source code.
+    """
+
+    class Clone(Redundancy):
+        """
+        This redundancy describes a code clone. Code clones are different pieces of
+        code in your codebase that are very similar.
+        """
+
+        MIN_CLONE_TOKEN = ("The number of tokens that have to be equal for it "
+                           "to be detected as a code clone.", int, 20)
+```
+
+This allows to show more information in results. If no custom message is given,
+the documentation of e.g. the Clone aspect class can be shown.
+
+If a bear does not implement all settings defined by the aspect, a debug
+message should warn the bear developer. If that setting is used in the
+configuration, the user shall be warned.
+
+## Implementation
+
+The implementation will have to be two phased as we should ideally deprecate
+the old way of using bears, keep it around for few releases and then phase it
+out and remove unneeded source code.
+
+### Gathering Metadata
+
+To get all bears for a set of aspects, all bears will have to be collected. It
+can then be filtered against it's metadata to get only bears that analyze the
+given aspects. (This information should be cached to improve the performance.)
+
+If multiple bears provide the same aspects a debug message should be emitted
+so developers consider removing the aspects from one of the bears and the lower
+bear with an alphabetic comparison will be run for that aspet.
```

Comment:
`aspect` typo

and about the alphabetical thing: how about choosing the bear that performs more aspects? This way, there'll be fewer bears to run, so performance will improve. I know that more aspects doesn't imply better quality results for any particular aspect, but neither does alphabetical sorting.

Or I could argue that we should choose the bear that solves fewer aspects so that it'll be more focussed on each aspect (may not necessarily be true).